### PR TITLE
Allows websocket calls throught shotgunlocalhost

### PR DESCRIPTION
**Pull request by geffrak**
In order to be able to send requests to Shotgun Create from a local tool, we
need to allow communications comming from localhost.

With these changes, a local process can talk to create using
"wss://shotgunlocalhost.com:PORT"

### CODE REVIEW

File: python/tk_desktop2/websockets/websockets_connection.py [hunk 1]
```diff
@@ -39,6 +40,10 @@ class WebsocketsConnection(object):
     _legacy_site_warning_displayed = False
     _legacy_user_warning_displayed = False
 
+    # localhost url regex matcher
+    # https://regex101.com/r/OeTZRx/4
```
Comment (jfboismenu) [inline python/tk_desktop2/websockets/websockets_connection.py]: ❤️

File: python/tk_desktop2/websockets/websockets_connection.py [hunk 2]
```diff
@@ -39,6 +40,10 @@ class WebsocketsConnection(object):
...
+    localhost_re = re.compile(r"^(?:[A-Za-z]+://)?(?:(?:shotgunlocalhost\.com)|(?:localhost))(?::\d+)?(?:/\S*)*$")
```
Comment (jfboismenu) [inline python/tk_desktop2/websockets/websockets_connection.py]: I'm a bit worried about the complexity of this regular expression. Does it need to be this complicated? Also, [REDOS](https://www.owasp.org/index.php/Regular_expression_Denial_of_Service_-_ReDoS).

I would expect the origin coming from a websocket API written by us or set by the browser to follow a pretty predictable pattern, no?

Right now it seems your regular expression accepts quite a few things that we probably don't want to even support in the real world like an origin without a protocol or a port number, which will never happen.

Comment (jfboismenu) [reply python/tk_desktop2/websockets/websockets_connection.py]: Naively, I would say we need to support something like [this](https://regex101.com/r/YRas6F/1)

Comment (geffrak) [reply python/tk_desktop2/websockets/websockets_connection.py]: I can live with your approach. I will adjust two things, 
1) For some reason, the origin is http on my python client so I think we need to support http as an origin.
2) The user can decide to change the port to something bigger than 9999

Comment (geffrak) [reply python/tk_desktop2/websockets/websockets_connection.py]: Is this better? 
https://regex101.com/r/YRas6F/2

There's no more cases where we can compute for ever.

Comment (jfboismenu) [reply python/tk_desktop2/websockets/websockets_connection.py]: I'm curious, why the need for the `(?:` [This](https://regex101.com/r/7n3JIp/1) would work as well. I'll be honest I've never had to use the `(?:` before, so I'm probably missing something here.

Comment (geffrak) [reply python/tk_desktop2/websockets/websockets_connection.py]: `( pattern )` is a capturing group
`( ?: pattern ) is a non-capturing group

Non-capturing groups are faster to evaluate.

Comment (geffrak) [reply python/tk_desktop2/websockets/websockets_connection.py]: What about this? https://regex101.com/r/7n3JIp/2 It catches http and https, localhost and shotgunlocalhost.com, port from 1 to 5 digits.

Comment (jfboismenu) [reply python/tk_desktop2/websockets/websockets_connection.py]: Ah sorry, I thought (?: was a bit more nefarious, I must have mixed it up with some other regex feature.

Yeah, let's keep your last proposition. Sorry for being pickly on something so simple.

File: python/tk_desktop2/websockets/websockets_connection.py [hunk 3]
```diff
@@ -39,6 +40,10 @@ class WebsocketsConnection(object):
     _legacy_site_warning_displayed = False
     _legacy_user_warning_displayed = False
 
+    # Pre-compiled shotgunlocalhost.com/localhost regex matcher
+    # You can play with it here: https://regex101.com/r/7n3JIp/3
+    localhost_re = re.compile(r"^https?://(?:shotgunlocalhost\.com|localhost):[0-9]{1,5}$")
```
Comment (manneohrstrom) [inline python/tk_desktop2/websockets/websockets_connection.py]: What about http? Please add a comment explaining why we are excluding non-encrypted local comms!

Comment (manneohrstrom) [reply python/tk_desktop2/websockets/websockets_connection.py]: If i understand this correctly, I think we should we also allow `127.0.0.1` as a valid local address. Again, if not, i think it is helpful to add a comment to explain why we have excluded it!

File: python/tk_desktop2/websockets/websockets_connection.py [hunk 4]
```diff
@@ -186,9 +191,9 @@ def _handle_server_id_request(self, message):
                 "Unexpected error while trying to retrieve the user id: "
                 "No user information was found in this request."
             )
-
-        # validate that the site of the request matches the sg create site
-        if self._origin_site != self._bundle.sgtk.shotgun_url:
+        
+        # validate that the site of the request matches shotgunlocalhost.com or the sg create site
+        if not WebsocketsConnection.localhost_re.match(self._origin_site) and self._origin_site != self._bundle.sgtk.shotgun_url:
```
Comment (manneohrstrom) [inline python/tk_desktop2/websockets/websockets_connection.py]: thinking about the support aspect here - when things don't work as expected and debugging is needed - will there be enough debug info emitted to be able to troubleshoot issues? Should we add debug to explain when we decide it's localhost vs external site for example?

File: python/tk_desktop2/websockets/websockets_connection.py (hunk 3 above)
Comment (geffrak) [reply python/tk_desktop2/websockets/websockets_connection.py]: https? mean http or https (look at the URL)
I can add 127.0.0.1 to the regex.

File: python/tk_desktop2/websockets/websockets_connection.py (hunk 4 above)
Comment (geffrak) [reply python/tk_desktop2/websockets/websockets_connection.py]: It's true that I had to inject a remote_pdb call to debug this part. Will add more debug logging.

**Pull request merged**